addon/services/key-manager.js

```diff
@@ -1,196 +1,159 @@
-  disableOnInput: false,
+export default Service.extend({
+  isDisabledOnInput: false, // Config option
 
   init() {
-    this._super(...arguments);
-    set(this, 'combos', []);
-    this._resetDownKeys();
-    this._registerConfig();
+    this.keydownMacros = [];
+    this.keyupMacros = [];
+    this._registerConfigOptions();
   },
 
-  register({keys, name, selector=$(document), downCallback, upCallback, priority=0, disableOnInput}) {
-    disableOnInput = disableOnInput || get(this, 'disableOnInput');
-
-    ['up', 'down'].forEach((direction) => {
-      const uid = get(this, 'uid');
-      const callback = direction === 'up' ? upCallback : downCallback;
-      const eventName = `key${direction}.${eventNamespace}.${name}.${uid}`;
-      const combo = Combo.create({
-        callback,
-        direction,
-        eventName,
-        keys,
-        name,
-        selector,
-        priority,
-        uid,
-        disableOnInput,
-      });
+  addMacro() {
+    const macroAttrs = this._mergeConfigDefaults(arguments[0]);
+    const macro = Macro.create();
+    macro.setup(macroAttrs);
 
-      get(this, 'combos').pushObject(combo);
-      selector.on(eventName, {
-        eventName,
-      }, run.bind(this, this.handler));
-      this.incrementProperty('uid');
-    });
-  },
+    const keyEvent = get(macro, 'keyEvent');
+    const macros = get(this, `${keyEvent}Macros`);
 
-  deregister({name}) {
-    const combos = get(this, 'combos');
-    const comboMatches = combos.filterBy('name', name);
+    const element = get(macro, 'element');
+    const hasMacrosWithSameElementAndKeyEvent = this._findMacroWithElementAndKeyEvent(element, keyEvent);
+    if (!hasMacrosWithSameElementAndKeyEvent) {
+      this._addEventListener(element, keyEvent);
+    }
 
-    comboMatches.forEach((comboMatch) => {
-      const {
-        eventName,
-        selector,
-      } = comboMatch;
+    macros.pushObject(macro);
 
-      selector.off(eventName);
-      combos.removeObject(comboMatch);
-    });
+    return macro;
   },
 
-  findMatchingCombo(event) {
-    try {
-      const { data, keyCode } = event;
-      const { eventName = null } = data;
+  _mergeConfigDefaults(attrs) {
+    const isDisabledOnInput = get(this, 'isDisabledOnInput');
+    return assign({ isDisabledOnInput }, attrs);
+  },
 
-      modifierKeys.forEach((key) => {
-        const prop = `${key}Key`;
-        set(this, prop, event[prop]);
-      });
+  _addEventListener(element, keyEvent) {
+    element.addEventListener(keyEvent, this);
+  },
 
-      if (!modifierKeyCodes.includes(keyCode)) {
-        get(this, 'downKeys').addObject(keyCode);
-      }
+  removeMacro({ element, keyEvent }) {
+    const macros = get(this, `${keyEvent}Macros`);
+    macros.removeObject(arguments[0]);
 
-      return this._findComboByName(eventName);
-    } finally {
-      this._clearExecutionKeys(event);
+    const hasMacrosWithSameElementAndKeyEvent = this._findMacroWithElementAndKeyEvent(element, keyEvent);
```

Comment:
Same thing - move this to a private method like an updated `_removeEventListener` since it is separate to the remove the object from registered macros.